### ISSUE TITLE
[bug] sqlite3: database disk image is malformed

### ISSUE THREAD

**Issue by kensanata** [superseriousbusiness/gotosocial#3499]
### Describe the bug with a clear and concise description of what the bug is.

When checking my database log, I see errors like the following:

> error dereferencing remote status … : enrichStatus: failed to dereference status author … : enrichAccount: error putting in database: sqlite3: database disk image is malformed (code=11 extended=11)

> 0xc0091c61e0: error processing: CreateAnnounce: error dereferencing announce: EnrichAnnounce: error fetching boost target … : enrichStatus: failed to dereference status author … : enrichAccount: error putting in database: sqlite3: database disk image is malformed (code=11 extended=11)

### What's your GoToSocial Version?

v17.1

### GoToSocial Arch

amd64

### What happened?

A few days ago I moved from v0.16.0 directly to v0.17.1. I did not install v0.17.0. I did not monitor my log files closely. I did not notice any strange behaviour. I use SQLite.

Yesterday figured I might create a second account for my blog on my own instance.
I tried to run `gotosocial admin account create` a few times, forgetting this or that parameter (see history below).
And then I noticed that the replies I saw scrolling by always ended in an error message.
In fact, there were more such error messages in my log file: "database disk image is malformed" 😱

I don't think my command-line fumbling is the cause for corruption, but just in case:

```
# history | grep account create
sudo -u gotosocial ./gotosocial admin account create --email '...' --username blog --password "..." --config-path config.yaml
sudo -u gotosocial ./gotosocial --config-path config.yaml admin account create --email "..." --username blog --password "..."
sudo -u gotosocial ./gotosocial admin account create --email "..." --username blog --password "..."
sudo -u gotosocial ./gotosocial admin account create --help
```

Example log:

```
# grep -i malformed /var/log/gotosocial/gotosocial.log
timestamp="30/10/2024 10:27:03.952" func=server.init.func1.Logger.13.1 level=ERROR latency="1.113342058s" userAgent="Mastodon/4.3.1 (http.rb/5.2.0; +https://chaos.social/)" method=POST statusCode=500 path=/users/alex/inbox clientIP=... pubKeyID=... errors="Error #01: AuthenticateFederatedRequest: error dereferencing account ...: enrichAccount: error putting in database: sqlite3: database disk image is malformed (code=11 extended=11)\n" requestID=pzyvzq4j04000p5drs1g msg="Internal Server Error: wrote 54B"
timestamp="30/10/2024 10:31:04.140" func=dereferencing.(*Dereferencer).DereferenceStatusDescendants level=ERROR username=alex status=... msg="error dereferencing remote status ...: enrichStatus: failed to dereference status author ...: enrichAccount: error putting in database: sqlite3: database disk image is malformed (code=11 extended=11)"
```

For comparison:

```
# wc -l /var/log/gotosocial/gotosocial.log
43054 /var/log/gotosocial/gotosocial.log
# grep -i malformed /var/log/gotosocial/gotosocial.log | wc -l
360
```

### What you expected to happen?

I don't expect database corruption.

### How to reproduce it?

No idea.

### Anything else we need to know?

There's something about these authors that's not working.

The code in `account.go` where the `error putting in database` string comes from:

```
// This is new, put it in the database.
err := d.state.DB.PutAccount(ctx, latestAcc)
if err != nil {
	return nil, nil, gtserror.Newf("error putting in database: %w", err)
}
```

I feel that this is where things are going wrong. Something about the accounts table.

I made an offline copy of the `sqlite.db` file. Sadly the `.recover` doesn't work:

```
$ sqlite3 sqlite.db ".recover" > data.sql
sql error: SQL logic error (1)
```

Not looking good! I'm going to try the dump.

```
sqlite3 sqlite.db ".dump" > data.sql
sqlite3 recovery.db < data.sql 2>&1 |tee recovery.log
```

Let's look at the log file and list the errors!

|Occurences|Type   |Error     |
|:---------|-------|----------|
|454|Runtime error|UNIQUE constraint failed: media_attachments.id |
|69|Runtime error|NOT NULL constraint failed: accounts.uri |
|2111|Parse error|no such table: sqlite_stat4 |
|1|Runtime error|NOT NULL constraint failed: conversations.thread_id|

The files I have available:

```
$ ls -lh data.sql recovery* sqlite.db 
-rw-r--r-- 1 alex alex    1.9G 30. Okt 10:23 data.sql
-rw-r--r-- 1 alex alex       0 30. Okt 10:39 recovery.db
-rw-r--r-- 1 alex alex    165K 30. Okt 10:39 recovery.log
-rw-r--r-- 2 alex lpadmin 9.8G 30. Okt 10:23 sqlite.db
```

The `recovery.db` is empty because `data.sql` ends with a rollback:

```
ROLLBACK; -- due to errors
```

Guessing that accounts are a problem:

```
# sqlite3 sqlite.db 
SQLite version 3.40.1 2022-12-28 14:03:47
Enter ".help" for usage hints.
sqlite> .schema accounts
CREATE TABLE IF NOT EXISTS "accounts" ("id" CHAR(26) NOT NULL, "created_at" timestamptz NOT NULL DEFAULT current_timestamp, "updated_at" timestamptz NOT NULL DEFAULT current_timestamp, "fetched_at" timestamptz, "username" VARCHAR NOT NULL, "domain" VARCHAR, "avatar_media_attachment_id" CHAR(26), "avatar_remote_url" VARCHAR, "header_media_attachment_id" CHAR(26), "header_remote_url" VARCHAR, "display_name" VARCHAR, "emojis" VARCHAR, "fields" VARCHAR, "fields_raw" VARCHAR, "note" VARCHAR, "note_raw" VARCHAR, "memorial" BOOLEAN DEFAULT false, "bot" BOOLEAN DEFAULT false, "locked" BOOLEAN DEFAULT true, "discoverable" BOOLEAN DEFAULT false, "uri" VARCHAR NOT NULL, "url" VARCHAR, "inbox_uri" VARCHAR, "shared_inbox_uri" VARCHAR, "outbox_uri" VARCHAR, "following_uri" VARCHAR, "followers_uri" VARCHAR, "featured_collection_uri" VARCHAR, "actor_type" VARCHAR NOT NULL, "private_key" VARCHAR, "public_key" VARCHAR NOT NULL, "public_key_uri" VARCHAR NOT NULL, "sensitized_at" timestamptz, "silenced_at" timestamptz, "suspended_at" timestamptz, "suspension_origin" CHAR(26), "public_key_expires_at" TIMESTAMPTZ, "moved_to_uri" VARCHAR, "also_known_as_uris" VARCHAR, "move_id" CHAR(26), PRIMARY KEY ("id"), UNIQUE ("id"), UNIQUE ("uri"), UNIQUE ("url"), UNIQUE ("inbox_uri"), UNIQUE ("outbox_uri"), UNIQUE ("following_uri"), UNIQUE ("followers_uri"), UNIQUE ("featured_collection_uri"), UNIQUE ("public_key_uri"), CONSTRAINT "usernamedomain" UNIQUE ("username", "domain"));
CREATE INDEX "accounts_domain_idx" ON "accounts" ("domain");
CREATE INDEX "accounts_paging_idx" ON accounts (COALESCE("domain", '') || '/@' || "username");
```

I'd be happy to provide more info.

**Comment by tsmethurst** [comment]
Hiya, could you read through this document and check if you're doing or have done any of the things listed in there, which might cause your database file to be corrupted? https://sqlite.org/howtocorrupt.html

Particularly, you should make sure you're not running on a network file system like NFS, or trying to back up your sqlite file in the middle of a transaction.

As for recovering your DB, have you checked this out? https://stackoverflow.com/a/15874884

**Comment by kensanata** [comment]
I'm looking at my shell history and while timestamps are missing, I think the order is essentially correct. But one possibility that I cannot rule out is that I did not wait long enough for gotosocial to release the database file. (Note that the history file has newer entries at the top.) From what I can tell, when I made the first `cp` I saw the hot files and copied them to the backup directory.  After getting the new release and unpacking it, however, I must have realised that the journal files are gone and used rsync to update the backup file. So far, so good.

```
systemctl status gotosocial
systemctl start gotosocial
emacs
tar xf gotosocial_0.17.1_linux_amd64.tar.gz
trash gotosocial_0.15.0_linux_amd64.tar.gz
rsync sqlite* backup/
ls *
wget https://github.com/superseriousbusiness/gotosocial/releases/download/v0.17.1/gotosocial_0.17.1_linux_amd64.tar.gz
cp sqlite* backup
mkdir backup
systemctl stop gotosocial
```

Now, here's a potential problem: At the time I didn't realize that my Monit would start gotosocial 10–15 minutes after going down. At the same time, my systemd service unit didn't allocate enough memory to gotosocial for the WASM compilation so it didn't finish starting up. I was assuming that migration was borked, so… and looking at it now, it seems that I disabled the service but didn't stop it before copying the backup file back! 🙈 I can't tell whether the service was stopped at that point and I felt it wasn't necessary to stop it and all I needed to do was prevent it from being started automatically, or whether this was the mistake I made.

```
systemctl enable gotosocial.service
cd /home/gotosocial
cp backup/sqlite.db .
systemctl disable gotosocial.service
systemctl status gotosocial.service
tail -f /var/log/gotosocial/gotosocial.log
less /var/log/monit.log
```

Thanks for that pointer to the recovery procedure. I will give it a try.

I guess with that the "bug" label no longer applies… Looks like a support issue, now.
